internal/corazawaf/rule.go

```diff
@@ -200,19 +257,44 @@ func (r *Rule) doEvaluate(phase types.RulePhase, tx *Transaction, cache map[tran
 			md.Data_ = r.LogData.Expand(tx)
 		}
 		matchedValues = append(matchedValues, md)
+		if multiphaseEvaluation {
+			*collectiveMatchedValues = append(*collectiveMatchedValues, md)
+		}
 		r.matchVariable(tx, md)
 	} else {
 		ecol := tx.ruleRemoveTargetByID[r.ID_]
 		for _, v := range r.variables {
-			// TODO(anuraaga): Support skipping variables based on phase for rule chains too.
-			if multiphaseEvaluation && !r.HasChain && r.ParentID_ == 0 {
+			if multiphaseEvaluation && r.ParentID_ == 0 && (!r.HasChain || phase >= r.chainMinPhase) {
 				min := minPhase(v.Variable)
 				// When multiphase evaluation is enabled, any variable is evaluated at its
-				// earliest possible phase, so we skip it if the rule refers to other phases
-				// to avoid double-evaluation.
-				if min != types.PhaseUnknown && min != phase {
-					continue
+				// earliest possible phase, so we make sure to skip in other phases.
+				if min != types.PhaseUnknown {
```

Comment:
Will this happen if multiphase isn't enabled? if nope then let's make explicit by adding `if multiphase evaluation` to the condition.

Reply:
It was inside an external `if multiphase`, but I made it clearer by moving all the logic into a `multiphaseSkipVariable` function. I'm just not sure if there might be possible stack/garbage collector concerns given the fact that this function will be called for every rule variable:
```
		for _, v := range r.variables {
			if multiphaseEvaluation && multiphaseSkipVariable(r, v.Variable, phase) {
				continue
			}
```